go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1263,6 +1274,40 @@ func (tsv *TabletServer) execDML(ctx context.Context, target *querypb.Target, qu
 	return int64(qr.RowsAffected), nil
 }
 
+// VStream streams VReplication events.
+func (tsv *TabletServer) VStream(ctx context.Context, target *querypb.Target, startPos mysql.Position, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error {
+	// This code is partially duplicated from execRequest. This is because
+	// is allowed even if the tablet is in non-serving state.
+	err := func() error {
+		tsv.mu.Lock()
+		defer tsv.mu.Unlock()
+
+		if target != nil {
+			// a valid target needs to be used
+			switch {
+			case target.Keyspace != tsv.target.Keyspace:
+				return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid keyspace %v", target.Keyspace)
+			case target.Shard != tsv.target.Shard:
+				return vterrors.Errorf(vtrpcpb.Code_INVALID_ARGUMENT, "invalid shard %v", target.Shard)
+			case target.TabletType != tsv.target.TabletType:
+				for _, otherType := range tsv.alsoAllow {
+					if target.TabletType == otherType {
+						return nil
+					}
+				}
+				return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "invalid tablet type: %v, want: %v or %v", target.TabletType, tsv.target.TabletType, tsv.alsoAllow)
+			}
+		} else if !tabletenv.IsLocalContext(ctx) {
```

Comment:
In the context of `VStream` does make sense to allow `Stream` request with nil targets? Would that be a table streaming to itself?

Reply:
It has to wait till first use. I've added an explanation.